background/services/name/resolvers/uns.ts

```diff
@@ -0,0 +1,169 @@
+import { ETHEREUM } from "../../../constants"
+import { sameNetwork } from "../../../networks"
+import { NameResolver } from "../name-resolver"
+
+/**
+ * Lookup a UNS domain name and fetch the owners address
+ */
+const lookupUNSDomain = async (domain: string) => {
+  const ALCHEMY_API_KEY = process.env.ALCHEMY_KEY
+  const response = await fetch(
+    `https://unstoppabledomains.g.alchemy.com/domains/${domain}`,
+    {
+      method: "GET",
+      headers: {
+        Authorization: `Bearer ${ALCHEMY_API_KEY?.trim()}`,
+      },
+    }
+  )
+  const data = await response.json()
+
+  return data
+}
+
+/**
+ * Reverse lookup an address and fetch it's corresponding UNS domain name
+ */
+const reverseLookupAddress = async (address: string) => {
+  const ALCHEMY_API_KEY = process.env.ALCHEMY_KEY
+  const response = await fetch(
+    `https://unstoppabledomains.g.alchemy.com/domains/?owners=${address}&sortBy=id&sortDirection=DESC`,
+    {
+      method: "GET",
+      headers: {
+        Authorization: `Bearer ${ALCHEMY_API_KEY?.trim()}`,
+      },
+    }
+  )
+  const data = await response.json()
+
+  return data
+}
+
+/**
+ * Check if a given string a valid UNS domain
+ */
+const IsValidUNSDomainName = (s: string): boolean => {
+  const trimmedString = s.trim()
+  let isValidUNSDomain = false
+
+  // Any valid domain name will have a period as part of the string
+  if (trimmedString.lastIndexOf(".") > 0) {
+    const domainExtension = trimmedString.slice(trimmedString.lastIndexOf("."))
+    const supportedUNSDomains = [
+      ".crypto",
+      ".coin",
+      ".wallet",
+      ".bitcoin",
+      ".888",
+      ".nft",
+      ".dao",
+      ".zil",
+      ".x",
+    ]
+
+    if (supportedUNSDomains.includes(domainExtension)) {
+      isValidUNSDomain = true
+    }
+  }
+  return isValidUNSDomain
+}
+
+export default function unsResolver(): NameResolver<"UNS"> {
+  return {
+    type: "UNS",
+    canAttemptNameResolution(): boolean {
+      return true
+    },
+    canAttemptAvatarResolution(): boolean {
+      return true
+    },
+    canAttemptAddressResolution({ name, network }: NameOnNetwork): boolean {
+      return sameNetwork(network, ETHEREUM) && IsValidUNSDomainName(name)
+    },
+
+    async lookUpAddressForName({
+      name,
+      network,
+    }: NameOnNetwork): Promise<AddressOnNetwork | undefined> {
+      // We try to resolve the name using unstoppable domains resolution
+      const data = await lookupUNSDomain(name)
+      const address = data.meta.owner
+
+      if (address === undefined || address === null) {
+        return undefined
+      }
+
+      return {
+        address,
+        network,
+      }
+    },
+    async lookUpAvatar(
+      addressOrNameOnNetwork: AddressOnNetwork | NameOnNetwork
+    ) {
+      const { network } = addressOrNameOnNetwork
+      const { address } =
+        "address" in addressOrNameOnNetwork
+          ? addressOrNameOnNetwork
+          : { address: undefined }
```

Comment:
Thoughts on adjusting this branch to attempt to resolve? Eg:

```suggestion
          : await this.lookUpAddressForName(addressOrNameOnNetwork) ?? { address: undefined }
```